src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/auth/securityconfigurer/CookieAndSessionCsrfTokenRepository.java

```diff
@@ -0,0 +1,153 @@
+ * This way we get the same token in session and csrf cookie
+ */
+public class CookieAndSessionCsrfTokenRepository implements CsrfTokenRepository {
+    public static final String DEFAULT_CSRF_COOKIE_NAME = "XSRF-TOKEN";
+    public static final String DEFAULT_CSRF_HEADER_NAME = "X-XSRF-TOKEN";
+    public static final String DEFAULT_CSRF_PARAMETER_NAME = "_csrf";
+
+    private static final String DEFAULT_CSRF_TOKEN_ATTR_NAME = CookieAndSessionCsrfTokenRepository.class
+            .getName().concat(".CSRF_TOKEN");
+
+    private final Method setHttpOnlyMethod;
+
+    private boolean cookieHttpOnly;
+    private String cookieName = DEFAULT_CSRF_COOKIE_NAME;
+    private String headerName = DEFAULT_CSRF_HEADER_NAME;
+    private String parameterName = DEFAULT_CSRF_PARAMETER_NAME;
+    private String sessionAttributeName = DEFAULT_CSRF_TOKEN_ATTR_NAME;
+
+    public CookieAndSessionCsrfTokenRepository() {
+        this.setHttpOnlyMethod = ReflectionUtils.findMethod(Cookie.class, "setHttpOnly", boolean.class);
+        if (this.setHttpOnlyMethod != null) {
+            this.cookieHttpOnly = true;
+        }
+    }
+
+    @Override
+    public CsrfToken generateToken(HttpServletRequest request) {
+        return new DefaultCsrfToken(this.headerName, this.parameterName,
+                createNewToken());
+    }
+
+    /**
+     * Saves a new token in session and adds that token in to a response cookie
+     * @param token token to be saved. The value of the token should not be empty or null
+     * @param request
+     * @param response
+     */
+    @Override
+    public void saveToken(CsrfToken token, HttpServletRequest request,
+            HttpServletResponse response) {
+        String tokenValue = token == null ? "" : token.getToken();
+        if (!StringUtils.isEmpty(tokenValue)) {
+            Cookie cookie = new Cookie(this.cookieName, tokenValue);
+            cookie.setSecure(request.isSecure());
+            cookie.setPath(this.getRequestContext(request));
+            if (token == null) {
+                cookie.setMaxAge(0);
+                HttpSession session = request.getSession(false);
+                if (session != null) {
+                    session.removeAttribute(this.sessionAttributeName);
+                }
+            } else {
+                cookie.setMaxAge(-1);
+                HttpSession session = request.getSession();
+                session.setAttribute(this.sessionAttributeName, token);
+            }
+            if (cookieHttpOnly && setHttpOnlyMethod != null) {
+                ReflectionUtils.invokeMethod(setHttpOnlyMethod, cookie, Boolean.TRUE);
+            }
+            response.addCookie(cookie);
+        }
+    }
+
+    /**
+     * The de facto token that gets loaded from the session
+     */
+    @Override
+    public CsrfToken loadToken(HttpServletRequest request) {
+        HttpSession session = request.getSession(false);
```

Comment:
We need our custom `CsrfValidationFilter` since the original `CsrfFilter` only verifies the one CsrfToken returned from `loadToken` here. So CsrfFilter only checks the value from **session attribute** and compares it to value from **request header**.

If we changed loadToken so that it 
- also loads value from **cookie**
- compares it to value from session attribute
- fails with `accessDeniedHandler.handle()` if those do not match

That way CsrfFilter + CookieAndSessionCsrfTokenRepository would be verifying that cookie == header == session attribute.

That way we could eliminate the custom CsrfValidationFilter and just use the original CsrFilter?

It is maybe a slight deviation of what you could expect from the players (repository stores and loads token, CsrfFilter validates token), but I did not see any hard rules in javadocs that we would be breaking, and it is also kind of logical; if repository stores tokens into two places, and loads only one value, it makes sense that it validates that value 1 == value 2 when loading one of the values.

Reply:
Basically all the validation logic was moved from the filter to this repository. If the validation fails, `loadToken` just returns `null` which causes 403 Forbidden.